### PR TITLE
Syslog parser works with any time format

### PR DESCRIPTION
**Pull request by ganmacs**
**Which issue(s) this PR fixes**: 
Fixes https://github.com/fluent/fluentd/issues/2856

**What this PR does / why we need it**: 

Setting `time_format`  in syslog hasn't worked well in some situations.
I made a change to work with all formats.  There is no performance degradation.

```rb
require 'benchmark/ips'
require 'fluent/env'
require 'fluent/engine'
require_relative '../lib/fluent/plugin/parser_syslog'

parser = Fluent::Plugin::SyslogParser.new
parser.configure(Fluent::Config::Element.new('ROOT', '', {'keep_time_key' => true}, []))
log = 'Feb 28 12:00:00 192.168.0.1 fluentd[11111]: [error] Syslog test'
parser2 = Fluent::Plugin::SyslogParser.new
parser2.configure(Fluent::Config::Element.new('ROOT', '', {'with_priority' => true, 'keep_time_key' => true}, []))
log2 = '<1>Feb 28 12:00:00 192.168.0.1 fluentd[11111]: [error] Syslog test'

Benchmark.ips do |x|
  x.report "regexp" do
    parser.parse(log) { |t, r| }
  end
  x.report "string" do
    parser.parse_rfc3164(log) { |t, r| }
  end
  x.report "regexp with pri" do
    parser2.parse(log2) { |t, r| }
  end
  x.report "string with pri" do
    parser2.parse_rfc3164(log2) { |t, r| }
  end
end
```

master branch

```
Warming up --------------------------------------
              regexp    22.312k i/100ms
              string    58.311k i/100ms
     regexp with pri    20.555k i/100ms
     string with pri    48.698k i/100ms
Calculating -------------------------------------
              regexp    248.764k (± 4.0%) i/s -      1.249M in   5.031343s
              string    669.950k (± 3.7%) i/s -      3.382M in   5.055510s
     regexp with pri    225.592k (± 3.5%) i/s -      1.131M in   5.017706s
     string with pri    554.281k (± 3.0%) i/s -      2.776M in   5.012618s
```

this branch
```
Warming up --------------------------------------
              regexp    24.435k i/100ms
              string    59.052k i/100ms
     regexp with pri    21.334k i/100ms
     string with pri    48.109k i/100ms
Calculating -------------------------------------
              regexp    262.246k (± 3.3%) i/s -      1.319M in   5.037310s
              string    644.082k (± 2.8%) i/s -      3.248M in   5.046689s
     regexp with pri    225.242k (± 3.5%) i/s -      1.131M in   5.026692s
     string with pri    510.475k (± 7.3%) i/s -      2.550M in   5.026552s
```

**Docs Changes**:

no need

**Release Note**: 

same as the title

### CODE REVIEW

File: lib/fluent/plugin/parser_syslog.rb [hunk 1]
```diff
@@ -36,8 +37,17 @@ class SyslogParser < Parser
       REGEXP_RFC5424_WITH_PRI = Regexp.new(<<~'EOS'.chomp % REGEXP_RFC5424, Regexp::MULTILINE)
         \A<(?<pri>[0-9]{1,3})\>[1-9]\d{0,2} %s\z
       EOS
+
       REGEXP_DETECT_RFC5424 = /^\<[0-9]{1,3}\>[1-9]\d{0,2}/
 
+      RFC3164_WITHOUT_TIME_AND_PRI_REGEXP = /(?<host>[^ ]*) (?<ident>[^ :\[]*)(?:\[(?<pid>[0-9]+)\])?(?:[^\:]*\:)? *(?<message>.*)$/
+      RFC3164_CAPTURES = ['host', 'ident', 'pid', 'message'].freeze
```
Comment (repeatedly) [inline lib/fluent/plugin/parser_syslog.rb]: `RFC3164_WITHOUT_TIME_AND_PRI_REGEXP.names.freeze`?

File: lib/fluent/plugin/parser_syslog.rb [hunk 2]
```diff
@@ -36,8 +37,17 @@ class SyslogParser < Parser
...
+      RFC3164_PRI_REGEXP = /^<(?<pri>[0-9]{1,3})>/
+
+      RFC5424_WITHOUT_TIME_AND_PRI_REGEXP = /(?<host>[!-~]{1,255}) (?<ident>[!-~]{1,48}) (?<pid>[!-~]{1,128}) (?<msgid>[!-~]{1,32}) (?<extradata>(?:\-|(?:\[.*?(?<!\\)\])+))(?: (?<message>.+))?\z/m
+      RFC5424_CAPTURES = ['host', 'ident', 'extradata', 'pid', 'msgid', 'message'].freeze
```
Comment (repeatedly) [inline lib/fluent/plugin/parser_syslog.rb]: ditto

File: lib/fluent/plugin/parser_syslog.rb [hunk 3]
```diff
@@ -103,51 +119,116 @@ def parse(text)
...
+            if @support_rfc5424_without_subseconds
+              log.trace(e)
+              @time_parser_rfc5424_without_subseconds.parse(time_str)
+            else
+              raise
+            end
+          end
+        end
+
+        record['time'] = time_str
```
Comment (repeatedly) [inline lib/fluent/plugin/parser_syslog.rb]: `record['time'] = time_str if @keep_time_key` is better than unnecessary `insert/delete` 'time' field.

File: lib/fluent/plugin/parser_syslog.rb [hunk 4]
```diff
@@ -103,51 +119,116 @@ def parse(text)
...
+          while text[i + 1] == ' '.freeze
+            sq = true
+            i += 1
+          end
+
+          i = text.index(' '.freeze, i + 1)
+        end
+
+        time_str = sq ? text.slice(idx, i - idx).squeeze(' ') : text.slice(idx, i - idx)
+        time = @mutex.synchronize { @time_parser.parse(time_str) }
```
Comment (repeatedly) [inline lib/fluent/plugin/parser_syslog.rb]: We can remove this mutex.

File: lib/fluent/plugin/parser_syslog.rb [hunk 5]
```diff
@@ -103,51 +119,116 @@ def parse(text)
...
+          while text[i + 1] == ' '.freeze
+            sq = true
+            i += 1
+          end
+
+          i = text.index(' '.freeze, i + 1)
+        }
+
+        time_str = sq ? text.slice(idx, i - idx).squeeze(' '.freeze) : text.slice(idx, i - idx)
+        time = @mutex.synchronize do
```
Comment (repeatedly) [inline lib/fluent/plugin/parser_syslog.rb]: Same with 3164

File: lib/fluent/plugin/parser_syslog.rb (hunk 4 above)
Comment (ganmacs) [reply lib/fluent/plugin/parser_syslog.rb]: I think [this code](https://github.com/fluent/fluentd/blob/5181d105d946a1fd71fdf147dbae931986d80647/lib/fluent/time.rb#L272-L275) can be a race condition. or I missed something?

File: lib/fluent/plugin/parser_syslog.rb (hunk 1 above)
Comment (ganmacs) [reply lib/fluent/plugin/parser_syslog.rb]: fixed https://github.com/fluent/fluentd/pull/2886/commits/f000aa2e891197d02df5e5f5a0bdb2457cbf47b3

File: lib/fluent/plugin/parser_syslog.rb (hunk 3 above)
Comment (ganmacs) [reply lib/fluent/plugin/parser_syslog.rb]: fixed https://github.com/fluent/fluentd/pull/2886/commits/5510b8f556b78cb9f268c01a1d08aef3bd7ed4cc

File: lib/fluent/plugin/parser_syslog.rb (hunk 4 above)
Comment (repeatedly) [reply lib/fluent/plugin/parser_syslog.rb]: Ah, I missed it. I forgot to improve this point, e.g. mutex should be moved to TimeParser and disable mutex feature for popular one-thread usecase.

**Comment by ganmacs** [comment]
@repeatedly ?

File: lib/fluent/plugin/parser_syslog.rb [hunk 6]
```diff
@@ -36,8 +37,17 @@ class SyslogParser < Parser
...
+
       REGEXP_DETECT_RFC5424 = /^\<[0-9]{1,3}\>[1-9]\d{0,2}/
 
+      RFC3164_WITHOUT_TIME_AND_PRI_REGEXP = /(?<host>[^ ]*) (?<ident>[^ :\[]*)(?:\[(?<pid>[0-9]+)\])?(?:[^\:]*\:)? *(?<message>.*)$/
+      RFC3164_CAPTURES = RFC3164_WITHOUT_TIME_AND_PRI_REGEXP.names.freeze
+      RFC3164_PRI_REGEXP = /^<(?<pri>[0-9]{1,3})>/
+
+      RFC5424_WITHOUT_TIME_AND_PRI_REGEXP = /(?<host>[!-~]{1,255}) (?<ident>[!-~]{1,48}) (?<pid>[!-~]{1,128}) (?<msgid>[!-~]{1,32}) (?<extradata>(?:\-|(?:\[.*?(?<!\\)\])+))(?: (?<message>.+))?\z/m
+      RFC5424_CAPTURES = RFC5424_WITHOUT_TIME_AND_PRI_REGEXP.names.freeze
+      RFC5424_PRI_REGEXP = /^<(?<pri>\d{1,3})>(?<version>\d\d{0,2})\s/
```
Comment (repeatedly) [inline lib/fluent/plugin/parser_syslog.rb]: Maybe, `version` is not needed?

Comment (ganmacs) [reply lib/fluent/plugin/parser_syslog.rb]: fixed https://github.com/fluent/fluentd/pull/2886/commits/8c6b8bda1e0f56f6ba79f068ed99ecd47827f87b

**Pull request merged**